scripts/evolution.py

```diff
@@ -0,0 +1,460 @@
+import sys
+import os
+import space
+import numpy as np
+import csv
+import json
+import copy
+import datetime
+from jsonschema import Draft4Validator, validators, exceptions
+from utility_functions import *
+from collections import defaultdict
+from scipy import stats
+import numpy.random as rd
+import matplotlib
+# matplotlib.use('TkAgg')
+from matplotlib import pyplot as plt
+
+
+def get_best_config(configs):
+    """
+    Returns the configuration with lowest Value among the given configurations
+    :param configs: list of configurations.
+    :return: dict, the best configuration.
+    """
+    leader = configs[0]
+    for c in configs:
+        if c['Value'] < leader['Value']:
+            leader = c
+    return leader
+
+
+def mutation(param_space, config, mutationrate, list=False):
+    """
+    Mutates given configuration. This overcomplicates the procedure. But since I might
+    change the functionality I leave it like this for now
+    :param param_space: space.Space(), will give us information about parameters
+    :param configs: list of configurations.
+    :param mutationrate: integer for how many parameters to mutate
+    :param list: boolean whether returning one or more alternative configs
+    :return: list of dicts, list of mutated configurations
+    """
+
+    parameter_object_list = param_space.get_input_parameters_objects()
+    rd_config = dict()
+    for name, obj in parameter_object_list.items():
+        x = obj.randomly_select()
+        # sucks that there are parameters with only one value
+        if x == config[name]:
+            x = obj.randomly_select()
+        rd_config[name] = x
+    parameter_names_list = param_space.get_input_parameters()
+    nbr_params = len(parameter_names_list)
+
+    configs = []
+    n_configs = nbr_params if list else 1
+
+    for _ in range(n_configs):
+        indices = rd.permutation(nbr_params)[:mutationrate]
+        for idx in indices:
+            mutation_param = parameter_names_list[idx]
+            # Should I do something if they are the same?
```

Comment:
Hmm, given the suggestion above it can only happen when a parameter only has one value or multiple identical values. When list == True it will not matter. When list == False the mutated config might be identical to the old config. If it is, I handle that in the evolutionary procedure by drawing a random config. 

As this scenario should be very rare I think it is okay as it is. Otherwise we single valued parameters could be removed from the loop.

Reply:
Given how messy these two issues became I admit that def mutation() not is the part of my work that I am the most proud of...

Reply:
I think the way it is now is good then. These two issues are more about corner cases causing issues that we want to avoid, the code is fine, don't worry :)